src/advanced-types/subtyping.md

```diff
@@ -1 +1,147 @@
+
+### Breaking backwards compatibility
+Lets demonstrate an upgrade that would NOT be backwards compatible. Lets upgrade to `V3`:
+```motoko
+{{#include _mo/subtyping.mo:i}}
+```
+
+We only changed the return type of the `get` function from `Nat` to `Int`. This change is NOT backwards compatible! The reason for this is that the new version of the public shared function `get` is not a *subtype* of the old version.  
+
+### Subtyping public shared functions
+Public shared functions in actors can have a subtype-supertype relationship. A public shared function `f1` could be a subtype of another public shared function `f2`, giving us the the relationship `f1 <: f2`.
+
+This means that we can use `f1` everywhere `f2` is expected, but not the other way around. But more importantly, we could safely upgrade from the supertype `f2` to the subtype `f1` without breaking backwards compatibility. 
+
+Lets look at an example where `f1 <: f2`:
+```motoko
+{{#include _mo/subtyping2.mo:a}}
+```
+
+Function `f1` is a subtype of `f2`. For this relationship to hold, two conditions must be satisfied:
+- The *return type* of `f1` must be a subtype of the return type of `f2`
+- The *argument(s)* of `f2` must be a subtype of the argument(s) of `f1`
+
+The return type `Nat` of function `f1` is a subtype of return type `Int` of function `f2`.   
+The argument `Nat` of function `f2` is a subtype of argument `Int` of function `f1`.
+
+> **NOTE**  
+> *There is an [exception](https://forum.dfinity.org/t/new-candid-version-and-catching-send-failures-motoko-updates/18410/3) for variant return types in new versions of [Candid](/internet-computer-programming-concepts/async-data/candid.html)*
+
+The types of `f1` and `f2` in Motoko are:
+```motoko
+{{#include _mo/subtyping2.mo:b}}
+```
+
+Upgrading a public shared function of type `F2` to a public shared function of type `F1` is considered a backwards compatible upgrade.
+
+Since `F1` is a subtype of `F2` and public shared functions are a [shared type](/internet-computer-programming-concepts/async-data/shared-types.html), we could use these types like this:
+```motoko
+{{#include _mo/subtyping2.mo:c}}
+```
+
+We are returning `f1` of type `F1` where a `F2` is expected. 
+
+### Upgrading public shared functions
+Suppose we have the following types:
+```motoko
+{{#include _mo/subtyping3.mo:a}}
+```
+
+Variant `Args` is a subtype of variant `ArgsV2`. Also, object type `ResultV2` is a subtype of object type `Result`.
+
+We may have a function that uses `Args` as the argument type and `Result` as the return type:
+```motoko
+{{#include _mo/subtyping3.mo:b}}
+```
+
+We could upgrade to a new version with types `ArgsV2` and `ResultV2` without breaking backwards compatibility, as long as we keep the same *function name*:
+```motoko
+{{#include _mo/subtyping4.mo:a}}
+```
+
+This is because `Args` is a *subtype* of `ArgsV2` and `Result` is a *supertype* of `ResultV2`.  
```

Comment:
```suggestion
This is because `Args` is a *subtype* of `ArgsV2` and `Result` is a *supertype* of `ResultV2`.  


Why is the compatible? The upgraded function can consume any arguments provided by old clients (it just accepts an additional variant from new clients but happily accepts the previous variants from old clients), and will produce all the fields required by old clients (but happens to provide an additional one for new clients).
```